spec/app/ics-029-fee-payment/README.md

```diff
@@ -0,0 +1,433 @@
+// Caller should provide the intended relayer address in case the fee is dependent on specific relayer(s).
+function GetReceiveFee(portID, channelID, sequence, relayer) Fee
+
+// Gets the fee expected for submitting AcknowledgePacket msg for the given packet
+// Caller should provide the intended relayer address in case the fee is dependent on specific relayer(s).
+function GetAckFee(portID, channelID, sequence, relayer) Fee
+
+// Gets the fee expected for submitting TimeoutPacket msg for the given packet
+// Caller should provide the intended relayer address in case the fee is dependent on specific relayer(s).
+function GetTimeoutFee(portID, channelID, sequence, relayer) Fee
+```
+
+Since different chains may have different representations for fungible tokens and this information is not being sent to other chains; this ICS does not specify a particular representation for the `Fee`. Each chain may choose its own representation, it is incumbent on relayers to interpret the Fee correctly.
+
+A default representation will have the following structure:
+
+```typescript
+interface Fee {
+  denom: string,
+  amount: uint256,
+}
+```
+
+### IBC Module Wrapper
+
+The fee middleware will implement its own ICS-26 callbacks that wrap the application-specific module callbacks as well as the ICS-4 handler functions called by the underlying application. This fee middleware will ensure that the counterparty module supports incentivization and will implement all fee-specific logic. It will then pass on the request to the embedded application module for further callback processing.
+
+In this way, custom fee-handling logic can be hooked up to the IBC packet flow logic without placing the code in the ICS-4 handlers or the application code. This is valuable since the ICS-4 handlers should only be concerned with correctness of core IBC (transport, authentication, and ordering), and the application handlers should not be handling fee logic that is universal amongst all other incentivized applications. In fact, a given application module should be able to be hooked up to any fee module with no further changes to the application itself.
+
+#### Fee Protocol Negotiation
+
+The fee middleware will negotiate its fee protocol version with the counterparty module by prepending its own version to the application version. 
+
+Channel Version: `fee_v{fee_protocol_version}:{application_version}`
+
+Ex: `fee_v1:ics20-1`
+
+The fee middleware's handshake callbacks ensure that both modules agree on compatible fee protocol version(s), and then pass the application-specific version string to the embedded application's handshake callbacks.
+
+#### Handshake Callbacks
+
+```typescript
+function onChanOpenInit(
+  order: ChannelOrder,
+  connectionHops: [Identifier],
+  portIdentifier: Identifier,
+  channelIdentifier: Identifier,
+  counterpartyPortIdentifier: Identifier,
+  counterpartyChannelIdentifier: Identifier,
+  version: string) {
```

Comment:
I don't understand exactly the implications on app developer/end user here re this change; do they need to know the version of fee middleware?Can someone clarify this a bit?More general, why do we need handshake between handlers if fee protocol does not require the other party to support it?

Reply:
No they will not, since the middleware will strip out its fee-specific version string before passing onto application. Thus, underlying application does not need to know it is being wrapped by middleware